sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/AdaptiveSparkPlanExec.scala

```diff
@@ -96,6 +96,7 @@ case class AdaptiveSparkPlanExec(
   @transient private val queryStageOptimizerRules: Seq[Rule[SparkPlan]] = Seq(
     PlanAdaptiveDynamicPruningFilters(this),
     ReuseAdaptiveSubquery(context.subqueryCache),
+    PrepareScans,
```

Comment:
@sunchao, per our design doc discussion, removing this explicit call causes test failures. I'll check what is going on tomorrow.

Reply:
Seems we don't support DPP + AQE atm (see https://issues.apache.org/jira/browse/SPARK-30186) - not sure if this is related.

Reply:
It's supported: https://issues.apache.org/jira/browse/SPARK-34168

Reply:
According to the design doc, output partitioning may change after runtime filter pushdown. But this is `queryStageOptimizerRules` and we can't add new shuffles here.

Reply:
I think @cloud-fan is right.

If I understand correctly, `AdaptiveSparkPlanExec` runs `queryStagePreparationRules` that adds shuffles on the input plan to get the initial physical plan. That plan is further analyzed and is split into query stages based on shuffles. Once Spark creates stages in a plan, it applies `queryStageOptimizerRules` and `postStageCreationRules` on each of them. A stage may have a join without shuffles (e.g. if the scan output partitioning is `SinglePartition`). If we dynamically filter the scan and it now reports 0 partitions, the stage execution would fail as the output partitioning is no longer valid.

The problem is that need to run filters after `PlanAdaptiveDynamicPruningFilters` which is part of the query optimizer rules in AQE.

Any thoughts, @cloud-fan @sunchao? Is it possible to run `queryStagePreparationRules` before executing the stage?

Reply:
While it would be easier to not allow data sources to change the partitioning during dynamic filtering, I am really trying to avoid that as it would be a substantial limitation. I mention this in the "Output Partitioning and Dynamic Filtering" section of the design doc. There are primarily two problems. First, if dynamic filtering is selective, this can lead to a lot of (almost) empty tasks being triggered hurting the performance. Second, it is hard to check whether the output partitioning matches. We can validate the number of partitions but the rest is hard.

Reply:
@JkSelf do you think it's possible to run `PlanAdaptiveDynamicPruningFilters` in `queryStagePreparationRules`?

Reply:
@cloud-fan  `PlanAdaptiveDynamicPruningFilters ` rule will add the DPP filter subquery and then the subquery will be executed when calling the `prepare()` method in `FileScan `operator. Here if adding the `PrepareScans `rule, the DPP filter subquery will be executed in this rule. I think there should be no problem that the DPP filter subquery is executed in the `PrepareScans `rule.  Please correct me if I don't understand your question.

Reply:
The problem is we can't add shuffles in the query optimization phase, and the options I can think of is:
1. doing DPP in the stage preparation phase.
2. improve the framework to allow adding shuffle in the query optimization phase.
3. do not apply DPP if more shuffles are introduced.

I think option 3 should be the simplest, but not sure if it's good enough for your use cases, @aokolnychyi

Reply:
also curious why `PlanAdaptiveDynamicPruningFilters` is in `queryStageOptimizerRules` but not `queryStagePreparationRules`. Any particular reason?

Reply:
It would be nice to move `PlanAdaptiveDynamicPruningFilters` and `PrepareScans` to prep rules in AQE just like I added `PrepareScans` to prep rules for non-AQE path. It is a bit tricky, though. `PlanAdaptiveDynamicPruningFilters` references the root node while prep rules are run immediately upon construction. Let me give it a try.

Reply:
I checked the source code a bit more, `PlanAdaptiveDynamicPruningFilters` at least needs to be run after `PlanAdaptiveSubqueries`.

Reply:
Thanks for looking into this, @cloud-fan!

I made dynamic filtering work in query prep rules but there are issues with dynamic subquery reuse. In short, if we want to reuse dynamic subqueries, we cannot execute them in prep rules. If we execute dynamic subqueries in prep rules, this will break `ReuseAdaptiveSubquery` and that one must be run in the query stage optimizer rules.

The more I think the more it feels like we won't be able to execute dynamic filters before running `EnsureRequirements` for the first time. That's why I am exploring the option of running prep rules one more time once a stage is created.

Does anybody have any thoughts on this? AQE currently runs prep rules only before creating stages.

Reply:
> Does anybody have any thoughts on this? AQE currently runs prep rules only before creating stages. Will it be too bad to call prep rules twice? Once before dynamic filtering and once after?

Technically, the plan after prep rules should reach a final status regarding exchanges. Once we expect some changes to exchanges in the plan (e.g. dynamic filtering change partitions of the stage), shouldn't we revert the stage and re-plan the adaptive plan? In some cases, is it possible that the stage will not exist after re-planning as the underlying partition is changed?

Reply:
Updated the approach to never introduce new shuffles.

Reply:
I am resolving this thread to not mislead other reviewers. It no longer applies.